fqdn: Make Rule UUIDs random instead of depending on the labels.

jrajahalme opened the pull request:
Rules may not have any labels and many rules may share the same
labels. Therefore a rule can't be uniquely identified by it's labels.

Generate a random rule UUID for each rule with "toFQDNs" so that we
can identify the same rule when needed for an update.

Unit tests are updated to not assume a stable UUID for a given set of
labels. Tests are also modified to make sure each rule has a unique
UUID regardless of it's labels, and that UUID is generated also if the
rule has no labels.

Fixes: #5890
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5897)
<!-- Reviewable:end -->

raybejjani reviewed (changes requested):
I'm sorry you had to look through the tests! they're quite obtuse. Thank you!
The UUIDs started as random, a long time ago, but I replaced them because we needed a consistent UUID when we replaced rules (i.e. on insert into the agent). If we change the toFQDNs list in an existing rule and `kubectl apply` the replace doesn't correctly remove names to be polled etc. but, more critically, we never replace the older rule data in `poller.allRules` because the new rule receives a new random UUID.

I prefer the random UUID, and this should sort this one case out:
```
diff --git a/daemon/policy.go b/daemon/policy.go
index a18ea4a6a..e76d5040f 100644
--- a/daemon/policy.go
+++ b/daemon/policy.go
@@ -211,6 +211,7 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
                        for _, r := range rules {
                                tmp := d.policy.SearchRLocked(r.Labels)
                                if len(tmp) > 0 {
+                                       d.dnsPoller.StopPollForDNSName(tmp)
                                        d.policy.DeleteByLabelsLocked(r.Labels)
                                }
                        }
@@ -218,6 +219,7 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
                if len(opts.ReplaceWithLabels) > 0 {
                        tmp := d.policy.SearchRLocked(opts.ReplaceWithLabels)
                        if len(tmp) > 0 {
+                               d.dnsPoller.StopPollForDNSName(tmp)
                                d.policy.DeleteByLabelsLocked(opts.ReplaceWithLabels)
                        }
                }
```

Unrelated to this PR, there is a race between the DNSPoller controller (`LookupUpdateDNS`) updating `poller.allRules` and adding new ones. I don't think that's happening here, and I've been meaning to replace allRules with a lookup into the policy repository, but in case you saw this behaviour this might be the reason.

Feel free to merge with the change above or some variant. I'm requesting changes to stop tgraf from optimistically merging this ;)

jrajahalme committed:
fqdn: Make Rule UUIDs random instead of depending on the labels.

Rules may not have any labels and many rules may share the same
labels. Therefore a rule can't be uniquely identified by it's labels.

Generate a random rule UUID for each rule with "toFQDNs" so that we
can identify the same rule when needed for an update.

Unit tests are updated to not assume a stable UUID for a given set of
labels. Tests are also modified to make sure each rule has a unique
UUID regardless of it's labels, and that UUID is generated also if the
rule has no labels.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

raybejjani reviewed (approved):
@jrajahalme exactly so. I've discussed adding hooks of some sort into the policy repo to catch the add/delete events. This would then simplify some of this logic with respect to both toFQDNs and toServices (we could reuse the part of the rule translator stuff in the k8s package).
